core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -377,45 +382,51 @@ private void removeContextListenerAndCancelDeadlineFuture() {
     class DeadlineExceededNotifyApplicationTimer implements Runnable {
       @Override
       public void run() {
-        final InsightBuilder insight = new InsightBuilder();
-        stream.appendTimeoutInsight(insight);
+        Status status = buildDeadlineExceededStatusWithRemainingNanos(remainingNanos);
+        deadlineCancellationSendToServerFuture = startDeadlineSendCancelToServerTimer(status);
+        executeCloseObserverInContext(observer, status);
+      }
+    }
 
-        long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
-        long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
+    return deadlineCancellationExecutor.schedule(
+        new LogExceptionRunnable(new DeadlineExceededNotifyApplicationTimer()),
+        remainingNanos,
+        TimeUnit.NANOSECONDS);
+  }
 
-        StringBuilder buf = new StringBuilder();
-        buf.append("deadline exceeded after ");
-        if (remainingNanos < 0) {
-          buf.append('-');
-        }
-        buf.append(seconds);
-        buf.append(String.format(".%09d", nanos));
-        buf.append("s. ");
-        buf.append(insight);
+  private Status buildDeadlineExceededStatusWithRemainingNanos(long remainingNanos) {
+    final InsightBuilder insight = new InsightBuilder();
+    stream.appendTimeoutInsight(insight);
 
-        final Status status = DEADLINE_EXCEEDED.augmentDescription(buf.toString());
-        deadlineCancellationSendToServerFuture =
-            startDeadlineSendCancelToServerTimer(status);
+    long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
+    long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
 
-        class CloseInContext extends ContextRunnable {
-          CloseInContext() {
-            super(context);
-          }
+    StringBuilder buf = new StringBuilder();
+    buf.append("deadline exceeded after ");
+    if (remainingNanos < 0) {
+      buf.append('-');
+    }
+    buf.append(seconds);
+    buf.append(String.format(".%09d", nanos));
+    buf.append("s. ");
+    buf.append(insight);
 
-          @Override
-          public void runInContext() {
-            closeObserver(observer, status, new Metadata());
-          }
-        }
+    return DEADLINE_EXCEEDED.augmentDescription(buf.toString());
+  }
 
-        callExecutor.execute(new CloseInContext());
+  private void executeCloseObserverInContext(final Listener<RespT> observer, final Status status) {
```

Comment:
There are other two call sites of `closeObserver()` at the start of `startInternal()` that can be simplified by calling this method.